### PR TITLE
A quick Swift5 property wrapper

### PR DESCRIPTION
**Pull request by Tyler-Keith-Thompson**
I scanned through current issues and noticed that somebody had played with this idea before and that it *might* be in 3.0. That being said I'd really love this functionality to be present and I believe we could get a quick win.

Different considerations:
If it'd be reasonable to ship a shared container with the library this property wrapper could be a little nicer:
```swift
@propertyWrapper
public struct DependencyInjected<Value> {
    let name:String?
    public init(wrappedValue value: Value?) {
        self.name = nil
    }
    public init(wrappedValue value: Value?, name:String?) {
        self.name = name
    }

    public lazy var wrappedValue: Value? = {
        Container.shared.resolve(Value.self, name: name)
    }()
}
```

This means that when using this in a class you don't need to add "= nil"
Example:
```swift
class SomeClass {
    @DependencyInjected var someType:SomeProtocol?
}
```

Instead of:
```swift
class SomeClass {
    @DependencyInjected var someType:SomeProtocol? = nil
}
```

As an aside one of my biggest issues with dependency injection frameworks is the "magic". I love that this one is relatively easy to follow and heavily makes use of swift generics. I believe utilizing a property wrapper like this can help get rid of some of the "magic" introduced with SwinjectStoryboard.

You can command + click on this property wrapper and follow through the implementation to understand what container is being used to resolve a dependency. 

NOTE: I am a big fan of this framework but have only started playing with it recently. It's possible if not likely that I've missed important functionality here.

### CODE REVIEW

File: Sources/DependencyInjectionPropertyWrapper.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//  Created by Tyler Thompson on 12/17/19.
+//
+
+import Foundation
+@propertyWrapper
+public struct DependencyInjected<Value> {
+    let name:String?
+    let container:Container
+    
+    public init(wrappedValue value: Value? = nil, container:Container, name:String? = nil) {
```
Comment (davidthorn) [inline Sources/DependencyInjectionPropertyWrapper.swift:14]: Please add documentation for each of the parameters. I understand that they are already pretty logical, but its better to have documentation for all public methods and their parameters.

File: Sources/DependencyInjectionPropertyWrapper.swift [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//  DependencyInjectionPropertyWrapper.swift
+//  Swinject
+//
+//  Created by Tyler Thompson on 12/17/19.
+//
+
+import Foundation
+@propertyWrapper
+public struct DependencyInjected<Value> {
+    let name:String?
```
Comment (davidthorn) [inline Sources/DependencyInjectionPropertyWrapper.swift:11]: Please declare the properties private and add documentation for each of them. Thanks

**Review by davidthorn — CHANGES_REQUESTED** [review]
Please correct the changes that I have requested first.

**Comment by tkohout** [comment]
I think this is a great feature to have, we should think about the API naming well because we will have to keep it for the future versions.

This is what @jakubvano has now in the WIP 3.0 branch now:

https://github.com/Swinject/Swinject/blob/bde873b53251a57edc069aaf098b22f18f203de4/Sources/2.0%20API/InstanceWrapper.swift

**Comment by Tyler-Keith-Thompson** [comment]
Alright, I've added the requested documentation and also brought it up to date with master, given it's been a while. I'm not quite clear on the API naming conventions to use. `Provider` (which is used in the linked blob) seems to be a different concept, if I'm understanding correctly.

**Pull request closed without merging**